### PR TITLE
openPMD: Support for mesh refinement

### PR DESCRIPTION
**Pull request by guj**
Implements the openPMD mesh-refinement extension: https://github.com/openPMD/openPMD-standard/pull/252

- [x] Field naming structure (ADIOS2):
```
 double  /data/1000/fields/j/x              {2048, 2048, 2048}
 double  /data/1000/fields/j/y              {2048, 2048, 2048}
 double  /data/1000/fields/j/z              {2048, 2048, 2048}
 double  /data/1000/fields/j_lvl1/x         {4096, 4096, 4096}
 double  /data/1000/fields/j_lvl1/y         {4096, 4096, 4096}
 double  /data/1000/fields/j_lvl1/z         {4096, 4096, 4096}
 double  /data/1000/fields/rho              {2048, 2048, 2048}
 double  /data/1000/fields/rho_lvl1         {4096, 4096, 4096}
```
(future: `_<P>` patch number for HDF5 as a fallback)

- [x] particles live on the finest level of each patch but get by default initialized on the coarse mesh

### CODE REVIEW

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 1]
```diff
@@ -265,8 +265,10 @@ void WarpXOpenPMDPlot::CloseStep (bool isBTD, bool isLastBTDFlush)
     // close BTD file only when isLastBTDFlush is true
     if (isBTD and !isLastBTDFlush) callClose = false;
     if (callClose) {
-        if (m_Series)
-            m_Series->iterations[m_CurrentStep].close();
+        if (!m_Series)
+            return;
+        auto iterations = m_Series->writeIterations();
+        iterations[m_CurrentStep].close();
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp]: Just to warn you: There might occur a tiny merge conflict once #1860 is merged here.

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 2]
```diff
@@ -281,23 +283,37 @@ WarpXOpenPMDPlot::Init (openPMD::Access access, const std::string& filePrefix, b
     std::string filepath = filePrefix;
     std::string const filename = GetFileName(filepath);
 
+    std::string useSteps = R"(
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp]: This looks unrelated to the MR functionality in this PR :)

File: Source/Diagnostics/WarpXOpenPMD.cpp (hunk 1 above)
Comment (ax3l) [reply Source/Diagnostics/WarpXOpenPMD.cpp]: Yes, the other PR was merged out. I re-added the functionality to this PR.

**Comment by ax3l** [comment]
After a little cleanup commit, I did more tests today (ADIOS2, HDF5) and the particles and fields look great, as well as particles transitioning between patches.

I am double-checking with @RevathiJambunathan that the back-transformed diagnostics is still working and after that I think it can go in :)

**Comment by RevathiJambunathan** [comment]
Thanks for this PR!

I tested this for Back Transformed diagnostics and get an error:
```
STEP 64 starts ...
STEP 64 ends. TIME = 6.814271277e-13 DT = 1.064729887e-14
Walltime = 19.13449687 s; This step = 0.184028665 s; Avg. per step = 0.2989765135 s
terminate called after throwing an instance of 'std::runtime_error'
  what():  [Series] Detected illegal access to iteration that has been closed previously.
SIGABRT
/usr/bin/addr2line: DWARF error: could not find variable specification at offset 240e1
/usr/bin/addr2line: DWARF error: could not find variable specification at offset 240e1
See Backtrace.0.0 file for details
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI COMMUNICATOR 3 DUP FROM 0
with errorcode 6.

```
Here is the input file

[inputs_3d_slice_reformatted.txt](https://github.com/ECP-WarpX/WarpX/files/6512894/inputs_3d_slice_reformatted.txt)

**Comment by guj** [comment]
> Thanks for this PR!
> 
> I tested this for Back Transformed diagnostics and get an error:
> 
> ```
> STEP 64 starts ...
> STEP 64 ends. TIME = 6.814271277e-13 DT = 1.064729887e-14
> Walltime = 19.13449687 s; This step = 0.184028665 s; Avg. per step = 0.2989765135 s
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  [Series] Detected illegal access to iteration that has been closed previously.
> SIGABRT
> /usr/bin/addr2line: DWARF error: could not find variable specification at offset 240e1
> /usr/bin/addr2line: DWARF error: could not find variable specification at offset 240e1
> See Backtrace.0.0 file for details
> --------------------------------------------------------------------------
> MPI_ABORT was invoked on rank 0 in communicator MPI COMMUNICATOR 3 DUP FROM 0
> with errorcode 6.
> ```
> 
> Here is the input file
> 
> [inputs_3d_slice_reformatted.txt](https://github.com/ECP-WarpX/WarpX/files/6512894/inputs_3d_slice_reformatted.txt)

Thanks I will take a look later today

**Comment by ax3l** [comment]
The crash is likely due to additional patch parts that should remain in  #1909 for the transition to group & variable based I/O options. I'll clean them out and re-test.

**Comment by guj** [comment]
> The crash is likely due to additional patch parts that should remain in #1909 for the transition to group & variable based I/O options. I'll clean them out and re-test.

Would be great see why 1909 is not passing the CI.

**Comment by ax3l** [comment]
I removed the un-related streaming API changes from this PR.
I performed @RevathiJambunathan's test again and BTD works again :)

Let's also add BTD with openPMD to CI in a PR.
